Standardize mpi-serial MPI function prototypes with MPICH

dqwu opened the pull request:
This PR introduces several updates to the mpi-serial library to
improve consistency and maintainability.

Changes:
* Align MPI function declarations with MPICH, including
  adjustments to const correctness and the use of arrays
  vs. pointers.
* Update corresponding function bodies to reflect the changes
  in declarations.
* Modify internal functions to ensure compatibility with
  updated MPI function signatures.
* Change buf in struct Req from int* to void* for better
  generality in handling user buffers.
* Remove the duplicate declaration of MPI_Get_library_version.

dqwu committed:
Align mpi-serial function prototypes with MPICH for consistency

The mpi-serial library supports most common MPI calls, but some
function prototypes differ from their MPICH counterparts. This
update ensures consistency by aligning function declarations
with those in MPICH 4.3.0 (see src/include/mpi_proto.h).

dqwu commented:
References:
https://fossies.org/linux/mpich/src/include/mpi_proto.h
https://github.com/E3SM-Project/scorpio/pull/639

Tests:
make tests: PASS
cime_develop test SMS_D_Ln9_Mmpi-serial.f19_g16_rx1.A (with latest SCORPIO master): PASS

jedwards4b reviewed (approved):
Thank you for this contribution.

jayeshkrishna commented on mpi.h:
```diff
@@ -335,24 +335,24 @@ extern int MPI_Group_intersection(MPI_Group group1, MPI_Group group2,
 extern MPI_Fint MPI_Group_c2f(MPI_Group group);
 
-extern int MPI_Init(int *argc, char **argv[]) ;
+extern int MPI_Init(int *argc, char ***argv) ;
```
MPI serial library does not have to have the same decl (argv[] vs *argv) as MPICH, although not an issue IMO

dqwu replied:
Yes, I just want to make it exactly the same as MPICH in this PR, even though this change is not strictly necessary.

jayeshkrishna commented on recv.c:
```diff
@@ -93,7 +93,7 @@ int MPI_Irecv(void *buf, int count, MPI_Datatype datatype,
     }
 
-  rreq->buf=buf;
+  rreq->buf=(int*)buf;
```
Is this type cast correct?

dqwu replied:
Yes, rreq->buf is int* type in the code.

jayeshkrishna replied:
Is the right change here to change Req->buf to void * instead?

dqwu replied:
> Is the right change here to change Req->buf to void * instead?

I did not know why buf is defined as int* in the struct Req. Need confirmation before we can change it to void* instead:
```
typedef struct
{
  pListitem listitem;        /* to allow Req to be removed from list */

  int *buf;
  int source;
  int tag;
  int complete;
  int count;
  MPI_Datatype type;  /* needed for test_simple_bindexed2 */
} Req;
```

jayeshkrishna replied:
Looks like the user buffer associated with the request to me. I would change it to void *

dqwu replied:
OK, I will update it.

jayeshkrishna commented on send.c:
```diff
@@ -82,7 +82,7 @@ int MPI_Isend(void *buf, int count, MPI_Datatype datatype,
     }
 
-  sreq->buf=buf;
+  sreq->buf=(int*)buf;
```
Can you confirm that this type cast is correct?

dqwu replied:
Yes, sreq->buf is int* type in the code. Without this type cast, there is a confirmed compiler warning.

jayeshkrishna commented on type.c:
```diff
@@ -491,14 +491,14 @@ int FC_FUNC( mpi_type_hvector, MPI_TYPE_HVECTOR )
 }
 
-int MPI_Type_hvector(int count, int blocklen, MPI_Aint stride,
+int MPI_Type_hvector(int count, int blocklength, MPI_Aint stride,
```
Why do we need this name change (I don't have any specific preference though)?

dqwu replied:
It is my own preference (I know it is not necessary).

dqwu replied:
MPICH version:
```
  371 int MPI_Type_hvector(int count, int blocklength, MPI_Aint stride, MPI_Datatype oldtype,
  372                      MPI_Datatype *newtype) MPICH_API_PUBLIC;
```

jayeshkrishna commented on type.c:
```diff
@@ -747,11 +747,11 @@ int FC_FUNC( mpi_type_ub, MPI_TYPE_UB )(int * type, MPI_Aint * ub, int * ierr)
 }
 
-int MPI_Type_ub(MPI_Datatype type, MPI_Aint * ub)
+int MPI_Type_ub(MPI_Datatype type, MPI_Aint *displacement)
```
Some of these changes are just changes in variable names. Do we need it?

dqwu replied:
I just want to make them exactly the same as MPICH in this PR, even though this change is not strictly necessary.

jayeshkrishna reviewed (commented):
Can you also include a summary of the changes in the PR (const correctness, arrays vs pointers etc)?

dqwu commented on mpi.h:
```diff
@@ -443,8 +443,7 @@ extern double MPI_Wtime(void);
  * Additional interfaces needed for compiling E3SM with gcc-14
  */
-extern int MPI_Get_Version(int *mpi_vers, int *mpi_subvers);
-extern int MPI_Get_library_version(char *version, int *resultlen);
```
The declaration of MPI_Get_library_version() is already present, so the duplicate declaration has been removed in this PR.

jayeshkrishna commented:
@dqwu : Either update the PR desc (PR commit message) or the change commit message to list out the changes in the PR (you can add the list to the current PR desc for ex). Also make the req buf type change in a separate commit

dqwu committed:
Change buf in struct Req from int* to void*

Updated buf to void* in struct Req for better generality in
handling user buffers.

Added an explicit cast in MPI_Isend() to suppress the warning
about discarded const qualifiers.

dqwu commented:
> Can you also include a summary of the changes in the PR (const correctness, arrays vs pointers etc)?

OK, I have updated the PR description.

dqwu commented:
@jedwards4b, is this PR good to go? Thanks!
